### PR TITLE
VBM-133: Add push notifications

### PR DESCRIPTION
**Pull request by natata7**
IMAGE 2022-09-07 20:32:29

Знімок екрана 2022-09-07 о 18 45 25

### CODE REVIEW

File: packages/mobile/android/app/build.gradle [hunk 1]
```diff
@@ -289,6 +290,12 @@ dependencies {
     implementation "androidx.swiperefreshlayout:swiperefreshlayout:1.0.0"
     implementation "androidx.core:core-splashscreen:1.0.0"
 
+    implementation project(':react-native-push-notification')
+
+    implementation "com.google.firebase:firebase-messaging:20.1.3"
+
+    implementation "com.google.firebase:firebase-analytics:17.3.0"
+
```
Comment (AleksandrTermenzhy) [inline packages/mobile/android/app/build.gradle:298]:
```suggestion
    implementation project(':react-native-push-notification')
    implementation "com.google.firebase:firebase-messaging:20.1.3"
    implementation "com.google.firebase:firebase-analytics:17.3.0"

```

File: packages/mobile/src/common/constants/push.ts [hunk 2]
```diff
@@ -0,0 +1,4 @@
+const PUSH_CHANNEL = 'vsebude_push_notifications';
+const PUSH_ICON = 'ic_notification';
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/common/constants/push.ts:2]:
```suggestion
const ANDROID_PN_DEFAULT_CHANNEL_ID = 'vsebude_push_notifications';
const ANDROID_PN_ICON_NAME = 'ic_notification';
```

File: packages/mobile/src/index.tsx [hunk 3]
```diff
@@ -8,11 +8,12 @@ import { Navigation } from '~/navigation/navigation';
 import { store } from '~/store/store';
 import { NavigationDarkTheme, NavigationTheme } from '~/config/config';
 import { useColorScheme } from '~/hooks/hooks';
-import { appService } from '~/services/services';
+import { appService, pushApi } from '~/services/services';
 import { i18 } from './localization/localization';
 
 i18();
 appService.init();
+pushApi.init();
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/index.tsx:16]: Place it inside appService.init()

File: packages/mobile/src/services/push-notifications/push-notifications.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import PushNotification from 'react-native-push-notification';
+import PushNotificationIOS from '@react-native-community/push-notification-ios';
+import { PUSH_CHANNEL, PUSH_ICON } from '~/common/constants/constants';
+
+type PushNotificationProps = {
+  title?: string;
+  body: string;
+};
+
+class PushController {
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/push-notifications/push-notifications.ts:10]:
```suggestion
class PushNotificationService {
```

File: packages/mobile/src/services/push-notifications/push-notifications.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
+import PushNotification from 'react-native-push-notification';
+import PushNotificationIOS from '@react-native-community/push-notification-ios';
+import { PUSH_CHANNEL, PUSH_ICON } from '~/common/constants/constants';
+
+type PushNotificationProps = {
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/push-notifications/push-notifications.ts:5]: There should be some type exported by library

File: packages/mobile/src/services/push-notifications/push-notifications.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    });
+  }
+
+  public async init() {
+    await this.createChannel();
+    this.configureNotifications();
+  }
+
+  private createChannel(): Promise<void> {
+    return new Promise((resolve) => {
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/push-notifications/push-notifications.ts:26]: Promise wrapper is redundant here

File: packages/mobile/src/services/push-notifications/push-notifications.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      PushNotification.createChannel(
+        {
+          channelId: PUSH_CHANNEL,
+          channelName: PUSH_CHANNEL,
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/push-notifications/push-notifications.ts:30]: Channel name should have some reasonable name as it's visible to user

File: packages/mobile/src/services/push-notifications/push-notifications.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+  private configureNotifications() {
+    PushNotification.configure({
+      onNotification: (notification) => {
+        this.showNotification({
+          body: notification.message.toString(),
+        });
+
+        // required on iOS only
+        notification.finish(PushNotificationIOS.FetchResult.NoData);
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/push-notifications/push-notifications.ts:45]:
```suggestion
        if (IS_IOS) {
          notification.finish(PushNotificationIOS.FetchResult.NoData);
        }
```

File: packages/mobile/src/services/push-notifications/push-notifications.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        () => resolve(),
+      );
+    });
+  }
+
+  private configureNotifications() {
+    PushNotification.configure({
+      onNotification: (notification) => {
+        this.showNotification({
+          body: notification.message.toString(),
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/push-notifications/push-notifications.ts:41]: What about title?

File: packages/mobile/src/services/services.ts [hunk 10]
```diff
@@ -47,6 +48,8 @@ const notification = new NotificationService();
 
 const appService = new AppService();
 
+const pushApi = new PushController();
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/services.ts:51]:
```suggestion
const pushNotification = new PushController();
```

File: packages/mobile/src/services/services.ts [hunk 11]
```diff
@@ -53,6 +54,9 @@ const personalInfoApi = new PersonalInfoApi({
 
 const appService = new AppService();
 
+const pushNotification = new PushNotificationService();
+pushNotification.init();
```
Comment (AleksandrTermenzhy) [inline packages/mobile/src/services/services.ts:58]: Only imports/exports are allowed to be here.
Place init() call into App service

**Review by AleksandrTermenzhy — APPROVED** [review]

**Review by VladimirBludov — APPROVED** [review]

**Review by seal323 — APPROVED** [review]

**Pull request merged**